### DISCUSSION TITLE
How to disable OLAP mode?

### DISCUSSION THREAD

**Discussion by baptisteArno** [planetscale/discussion#408] [Q&A] [answered]
I have SELECT queries that can run for more than 30s.

I'm guessing that's because I have OLAP mode activated? How can I return to default mode?

**Comment by fooforge** [comment, marked as answer]
Hi @baptisteArno,

OLTP is the default workload mode and a new session will always start in OLTP mode. You cannot change the workload mode globally, but only switch to OLAP by manually executing `set workload='olap';`, or by adding it to the list of statements your database driver executes when it invokes the connection.

I know that we have a 30 seconds timeout for `SELECT`'s documented [here](https://planetscale.com/docs/reference/planetscale-system-limits), but I'm not too sure that's accurate anymore. We have increased that timeout to 900 seconds at some point. I'll poke around internally and read up on the code to see where the timeout is at now. Either way, no need to worry about the default workload mode.

**Reply by baptisteArno** [reply]
Alright, thank you 🙏

**Reply by ctc24** [reply]
Hi @fooforge,

> I know that we have a 30 seconds timeout for `SELECT`'s documented [here](https://planetscale.com/docs/reference/planetscale-system-limits), but I'm not too sure that's accurate anymore. We have increased that timeout to 900 seconds at some point. I'll poke around internally and read up on the code to see where the timeout is at now. Either way, no need to worry about the default workload mode.

Any word on what the current `SELECT` timeout value is? We're running in OLAP mode and are experiencing timeouts at around the 30s mark. This doesn't happen with other MySQL databases we work with, so we're wondering whether this might be due to an internal PlanetScale setting.

Any guidance you have would be helpful. Thank you!

**Reply by fooforge** [reply]
@ctc24 I have run some _quick_ tests today.

When running in our default OLTP workload mode, single `SELECT` queries time out at the 900 seconds mark and transactions after 20 seconds. After changing the workload mode to OLAP these restrictions are lifted.

Please note that you will only be able to change the workload mode on a per-session basis, meaning you will need to execute `SET workload='olap';` directly before executing your long-running query and within the same session. New sessions always default to the OLTP workload mode and there is no global toggle for it at this time.

I suspect, the reason why you're seeing your `SELECT` query timing out at the 20 seconds mark, is because you or your ORM wraps the query inside a transaction.

I'm going to file a pull request later today to update the per-query `SELECT` timeout value in [our system limits](https://https://planetscale.com/docs/reference/planetscale-system-limits). The recommendations and warnings around using OLAP mode still apply.

### PlanetScale in default (OLTP) mode:

```sql
connect-test/main> SELECT SLEEP(900);
ERROR 1105 (HY000): target: connect-test.-.primary: vttablet: rpc error: code = Canceled desc = (errno 2013) due to context deadline exceeded, elapsed time: 14m59.999365441s, killing query ID 65532 (CallerID: planetscale-admin)
connect-test/main> SELECT SLEEP(899);
+------------+
| SLEEP(899) |
+------------+
|          0 |
+------------+
connect-test/main> START TRANSACTION; SELECT SLEEP(20); COMMIT;
ERROR 1105 (HY000): target: connect-test.-.primary: vttablet: (errno 2013) due to context deadline exceeded, elapsed time: 19.999864427s, killing query ID 68345 (CallerID: planetscale-admin)
ERROR 1105 (HY000): target: connect-test.-.primary: vttablet: rpc error: code = Aborted desc = transaction 1681792480304955294: ended at 2023-04-20 12:35:25.267 UTC (unlocked closed connection) (CallerID: planetscale-admin)
connect-test/main> START TRANSACTION; SELECT SLEEP(19); COMMIT;
+-----------+
| SLEEP(19) |
+-----------+
|         0 |
+-----------+
```

### PlanetScale running in OLAP mode

```sql
connect-test/main> SET workload='olap';
connect-test/main> SELECT SLEEP(901);
+------------+
| SLEEP(901) |
+------------+
|          0 |
+------------+
connect-test/main> START TRANSACTION; SELECT SLEEP(900); COMMIT;
+------------+
| SLEEP(900) |
+------------+
|          0 |
+------------+
```

**Reply by ctc24** [reply]
@fooforge Thanks for the update and the thorough write-up! One quick follow up question for you: is there any way to update the transaction timeout while in OLAP mode?

We effectively have to rely on transactions to set our connection to OLAP mode. We connect to PlanetScale from a GoLang application. Go doesn't provide any guarantees around the fact that a connection will be reused between two different queries, unless they're wrapped in the same transaction. As such, the only way we can guarantee that a given query is running in OLAP mode is to do the following:
```
START TRANSACTION;
SET workload='olap';
SELECT ...;
COMMIT;
```

The last code snippet you posted above seems to suggest that you're able to run transactions that are longer than 30s in OLAP mode, but that hasn't been our experience. I'm able to reproduce the timeout both from our own Go application as well as from a vanilla MySQL shell.

Thanks again for your help!

**Reply by fooforge** [reply]
@ctc24 Sounds like you're trying to change the workload mode after the transaction has already been initiated. This means, however, that the transaction will still be dependent on the configured system limits in our default workload mode. This is a session variable and it's not possible to change it per transaction.

Subsequent transactions will work just fine afterwards since these then adhere to the settings of the OLAP workload mode.

```sql
connect-test/main> START TRANSACTION; SET workload='olap'; SELECT SLEEP(30); COMMIT;
ERROR 1105 (HY000): rpc error: code = Unknown desc = target: connect-test.-.primary: vttablet: (errno 2013) due to context deadline exceeded, elapsed time: 29.99989044s, killing query ID 71504 (CallerID: planetscale-admin)
connect-test/main> START TRANSACTION; SELECT SLEEP(30); COMMIT;
+-----------+
| SLEEP(30) |
+-----------+
|         0 |
+-----------+
```

You will need to execute the workload mode statement before you run your first transaction. I'm not too familiar with Go, but I would likely set up two separate connectors, one for OLTP and one for OLAP. For the latter, I'd execute a `SET workload='olap';` each time the application initiates a new connection.

```sql
SET workload='olap';
START TRANSACTION;
SELECT SLEEP(30);
COMMIT;
```

If that didn't help, mind sharing a real world example with some sample code?

**Reply by ctc24** [reply]
Thanks @fooforge. We got past this issue by restricting Go to a single DB connection for PlanetScale. This allows us to set OLAP mode when we initialize the connection and then be confident that this same connection is used for the query.

What we're now running into is that the query seems to run fine (no errors are thrown), but it returns a single row filled with `NULL` values. This is unexpected behavior: we know there is data in this table that should be returned. We have reason to believe that our code around this is working fine: it works in non OLAP mode, works for tables that have small result sets (eg a query that returns only 100 rows works fine), and works on other MySQL-protocol databases like ClickHouse and vanilla MySQL without issue.

This leads us to believe that there is some bug in how PlanetScale's OLAP mode handles longer-running / larger result-set queries.

**Reply by fooforge** [reply]
@ctc24 Mind opening a ticket with us at https://support.planetscale.com/ with the exact query and the different results that you see? We'll look into it then.

**Reply by ctc24** [reply]
Just pinged the ticket with the same details as above
